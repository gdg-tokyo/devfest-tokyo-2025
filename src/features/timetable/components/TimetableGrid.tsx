import React from 'react'
import SessionCard from './SessionCard'
import { OldSession } from '@/lib/data-parser'

interface TimetableGridProps {
  sessions: OldSession[]
  filters: { levels: string[]; keyword: string }
}

const minutesPerSlot = 10 // Granularity of time slots

const getTrackColor = (track: string) => {
  switch (track) {
    case 'Track A':
      return 'bg-gdg-pastel-red'
    case 'Track B':
      return 'bg-gdg-pastel-blue'
    case 'Track C':
      return 'bg-gdg-pastel-green'
    case 'Track D':
      return 'bg-gdg-pastel-yellow'
    default:
      return 'bg-gray-200'
  }
}

const TimetableGrid: React.FC<TimetableGridProps> = ({ sessions, filters }) => {
  // Dynamically extract unique tracks from sessions and sort them
  const tracks = Array.from(
    new Set(sessions.map((session) => session.track))
  ).sort()

  // If there are no sessions, provide a default set of tracks to ensure grid structure
  const displayTracks =
    tracks.length > 0 ? tracks : ['Track A', 'Track B', 'Track C', 'Track D']

  // Determine the overall time range
  const allStartTimes = sessions.map((s) => s.time_start)
  const allEndTimes = sessions.map((s) => s.time_end)

  const earliestTime =
    allStartTimes.length > 0 ? allStartTimes.sort()[0] : '09:00' // Default if no sessions
  const latestTime =
    allEndTimes.length > 0 ? allEndTimes.sort().reverse()[0] : '18:00' // Default if no sessions

  // Generate all 10-minute time slots between earliest and latest time
  const generateTimeSlots = (start: string, end: string): string[] => {
    const slots: string[] = []
    let [startHour, startMinute] = start.split(':').map(Number)
    let [endHour, endMinute] = end.split(':').map(Number)

    let currentTime = new Date()
    currentTime.setHours(startHour, startMinute, 0, 0)

    const endTime = new Date()
    endTime.setHours(endHour, endMinute, 0, 0)

    while (currentTime <= endTime) {
      slots.push(currentTime.toTimeString().substring(0, 5))
      currentTime.setMinutes(currentTime.getMinutes() + minutesPerSlot)
    }
    return slots
  }

  const allTimeSlots = generateTimeSlots(earliestTime, latestTime)

  const getRowStart = (time: string): number => {
    const index = allTimeSlots.indexOf(time)
    return index !== -1 ? index + 2 : 1 // +2 because first row is for track headers, and grid starts at 1
  }

  const getRowSpan = (startTime: string, endTime: string): number => {
    const [startHour, startMinute] = startTime.split(':').map(Number)
    const [endHour, endMinute] = endTime.split(':').map(Number)

    const startDate = new Date()
    startDate.setHours(startHour, startMinute, 0, 0)

    const endDate = new Date()
    endDate.setHours(endHour, endMinute, 0, 0)

    const durationMs = endDate.getTime() - startDate.getTime()
    const durationMinutes = durationMs / (1000 * 60)

    return Math.max(1, Math.ceil(durationMinutes / minutesPerSlot))
  }

  const filterSession = (session: Session): boolean => {
    const levelMatch =
      filters.levels.length === 0 ||
      (session.level && session.level.some((l) => filters.levels.includes(l)))
    const keywordMatch =
      !filters.keyword ||
      session.title.toLowerCase().includes(filters.keyword.toLowerCase()) ||
      session.longDescription
        .toLowerCase()
        .includes(filters.keyword.toLowerCase()) ||
      session.talks
        .flatMap((talk) => talk.speakers.map((speaker) => speaker.name))
        .some((name) =>
          name.toLowerCase().includes(filters.keyword.toLowerCase())
        )
    return levelMatch && keywordMatch
  }

  return (
    <div className="overflow-x-auto">
      <div
        className="grid gap-2 p-2"
        style={{
          gridTemplateColumns: `120px repeat(${displayTracks.length}, 1fr)`,
          gridTemplateRows: `[tracks] auto repeat(${allTimeSlots.length}, minmax(40px, auto))`,
        }}
      >
        {/* Corner for time/track headers */}
        <div className="sticky left-0 top-0 p-3 z-10 subhead text-sm text-gray-600">
          Time
        </div>

        {/* Track Headers */}
        {displayTracks.map((track, index) => (
          <div
            key={track}
            className={`p-3 rounded-lg border-2 border-gray-800 text-center sticky top-0 z-10 ${getTrackColor(track)}`}
            style={{ gridColumn: index + 2, gridRow: 1 }}
          >
            <h5>{track}</h5>
          </div>
        ))}

        {/* Time labels */}
        {allTimeSlots.map((time, index) => (
          <div
            key={time}
            className="sticky left-0 p-2 border-r border-gray-300 font-roboto-mono text-sm text-black-02 flex items-center justify-center z-10"
            style={{ gridRow: index + 2, gridColumn: 1 }}
          >
            {time}
          </div>
        ))}

        {/* Sessions */}
        {sessions.map((session) => {
          const trackIndex = displayTracks.indexOf(session.track)
          // Only render sessions that belong to one of the displayed tracks
          if (trackIndex === -1) return null

          const rowStart = getRowStart(session.time_start)
          const rowSpan = getRowSpan(session.time_start, session.time_end)
          const isGrayedOut = !filterSession(session)

          return (
            <div
              key={session.id}
              className={`font-google-sans text-black-02 ${isGrayedOut ? 'opacity-30' : ''}`}
              style={{
                gridRow: `${rowStart} / span ${rowSpan}`,
                gridColumn: trackIndex + 2,
                zIndex: 5, // Ensure session cards are above grid lines
              }}
            >
              <SessionCard session={session} isGrayedOut={isGrayedOut} />
            </div>
          )
        })}
      </div>
    </div>
  )
}

export default TimetableGrid
