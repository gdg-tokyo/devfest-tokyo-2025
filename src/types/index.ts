export interface OldSpeaker {
  id: string
  name: string
  job: string
  bio: string
  twitter_handle: string
}

export interface OldTalk {
  id: string
  title: string
  abstract: string
  speaker_ids: string[]
  tech_tags: string[]
  level?: 'Beginner' | 'Intermediate' | 'Advanced'
  perspective?: 'Introduction' | 'Experience' | 'Challenge'
  is_keynote: boolean
}

export interface OldSession {
  id: string
  talk_ids: string[]
  track: string
  time_start: string
  time_end: string
  room: string
  title: string
  level: string[]
  tech_tags: string[]
  description: string
}

export interface Speaker {
  id: string
  name: string
  bio: string
  photo_url: string
  job: string
  twitter_handle: string
}

export interface Talk {
  id: string
  title: string
  abstract: string
  time_start: string
  time_end: string
  track: string
  speaker_ids: string[]
  tech_tags: string[]
  level: ('Beginner' | 'Intermediate' | 'Advanced')[]
  perspective: ('Introduction' | 'Experience' | 'Challenge')[]
}
