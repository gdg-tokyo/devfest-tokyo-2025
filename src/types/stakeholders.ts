export interface Stakeholder {
  name: string
  logoUrl: string
  type: 'organizer' | 'partner'
  link: string
}
