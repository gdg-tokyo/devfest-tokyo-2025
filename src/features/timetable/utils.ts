import { Session, Talk, Speaker } from '@/types'
import { getTalks, getSpeakers } from '@/lib/data-parser'

const minutesPerSlot = 10 // Granularity of time slots

export const generateTimeSlots = (sessions: Session[]): string[] => {
  if (sessions.length === 0) {
    // Default times, still 30 min intervals for default
    return [
      '09:00',
      '10:00',
      '11:00',
      '12:00',
      '13:00',
      '14:00',
      '15:00',
      '16:00',
      '17:00',
      '18:00',
    ]
  }

  const allTimes = sessions.flatMap((s) => [s.time_start, s.time_end])

  const dateTimes = allTimes.map((timeStr) => {
    const [h, m] = timeStr.split(':').map(Number)
    const d = new Date()
    d.setHours(h, m, 0, 0)
    return d
  })

  const earliestTime = new Date(
    Math.min(...dateTimes.map((dt) => dt.getTime()))
  )
  const latestTime = new Date(Math.max(...dateTimes.map((dt) => dt.getTime())))

  // Round down earliest time to the nearest 10-minute interval
  earliestTime.setMinutes(Math.floor(earliestTime.getMinutes() / 10) * 10)
  earliestTime.setSeconds(0)
  earliestTime.setMilliseconds(0)

  // Round up latest time to the nearest 10-minute interval
  latestTime.setMinutes(Math.ceil(latestTime.getMinutes() / 10) * 10)
  latestTime.setSeconds(0)
  latestTime.setMilliseconds(0)

  const finalTimeSlots: string[] = []
  let currentTime = new Date(earliestTime.getTime())

  while (currentTime.getTime() <= latestTime.getTime()) {
    finalTimeSlots.push(currentTime.toTimeString().substring(0, 5))
    currentTime.setMinutes(currentTime.getMinutes() + 10) // 10-minute step
  }

  return finalTimeSlots
}

export const getRowStart = (time: string, allTimeSlots: string[]): number => {
  const index = allTimeSlots.indexOf(time)
  return index !== -1 ? index + 2 : 1 // +2 because first row is for track headers, and grid starts at 1
}

export const getRowEndLine = (
  startTime: string,
  endTime: string,
  allTimeSlots: string[]
): number => {
  const startIndex = allTimeSlots.indexOf(startTime)
  if (startIndex === -1) {
    // Fallback to minute-based calculation for span if start time is not found
    const [startHour, startMinute] = startTime.split(':').map(Number)
    const [endHour, endMinute] = endTime.split(':').map(Number)

    const startDate = new Date()
    startDate.setHours(startHour, startMinute, 0, 0)

    const endDate = new Date()
    endDate.setHours(endHour, endMinute, 0, 0)

    const durationMs = endDate.getTime() - startDate.getTime()
    const durationMinutes = durationMs / (1000 * 60)

    // Returns span, not end line, in this fallback case
    return Math.max(1, Math.ceil(durationMinutes / 10))
  }

  const [endHour, endMinute] = endTime.split(':').map(Number)
  const sessionEndTime = new Date()
  sessionEndTime.setHours(endHour, endMinute, 0, 0)

  let endLineIndex = startIndex // Initialize with start index
  for (let i = startIndex; i < allTimeSlots.length; i++) {
    const [slotHour, slotMinute] = allTimeSlots[i].split(':').map(Number)
    const slotTime = new Date()
    slotTime.setHours(slotHour, slotMinute, 0, 0)

    if (slotTime.getTime() < sessionEndTime.getTime()) {
      endLineIndex = i + 1 // The line *after* this slot
    } else if (slotTime.getTime() === sessionEndTime.getTime()) {
      endLineIndex = i // The line *at* this slot
      break
    } else {
      break // sessionEndTime is before current slotTime
    }
  }
  // +2 because first row is for track headers, and grid starts at 1
  return endLineIndex + 2
}

export const getTrackColor = (track: string) => {
  switch (track) {
    case 'Track A':
      return 'bg-gdg-pastel-red'
    case 'Track B':
      return 'bg-gdg-pastel-blue'
    case 'Track C':
      return 'bg-gdg-pastel-green'
    case 'Track D':
      return 'bg-gdg-pastel-yellow'
    default:
      return 'bg-gray-200'
  }
}

export const filterSession = (
  session: Session,
  filters: { levels: string[]; keyword: string }
): boolean => {
  const allRawTalks = getTalks()
  const allSpeakers = getSpeakers()

  const talksMap = new Map<string, Talk>(
    allRawTalks.map((talk) => [talk.id, talk])
  )
  const speakersMap = new Map<string, Speaker>(
    allSpeakers.map((speaker) => [speaker.id, speaker])
  )

  const levelMatch =
    filters.levels.length === 0 ||
    (session.level && session.level.some((l) => filters.levels.includes(l)))

  const sessionTalks = session.talk_ids
    .map((talkId) => talksMap.get(talkId))
    .filter((talk): talk is Talk => talk !== undefined)

  const speakerNames = sessionTalks.flatMap((talk) =>
    talk.speaker_ids
      .map((speakerId) => speakersMap.get(speakerId)?.name)
      .filter((name): name is string => name !== undefined)
  )

  const keywordMatch =
    !filters.keyword ||
    session.title.toLowerCase().includes(filters.keyword.toLowerCase()) ||
    session.description.toLowerCase().includes(filters.keyword.toLowerCase()) ||
    speakerNames.some((name) =>
      name.toLowerCase().includes(filters.keyword.toLowerCase())
    ) ||
    sessionTalks.some((talk) =>
      talk.abstract.toLowerCase().includes(filters.keyword.toLowerCase())
    )
  return levelMatch && keywordMatch
}
