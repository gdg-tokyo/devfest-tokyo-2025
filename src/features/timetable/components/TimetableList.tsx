'use client'

import { Session } from '@/types'
import React from 'react'
import { getTrackColor } from '../utils' // Import getTrackColor
import SessionCard from './SessionCard'

interface TimetableListProps {
  sessions: Session[]
  filters: { levels: string[]; keyword: string }
  allTimeSlots: string[]
  filterSession: (
    session: Session,
    filters: { levels: string[]; keyword: string }
  ) => boolean
}

const TimetableList: React.FC<TimetableListProps> = ({
  sessions,
  filters,
  allTimeSlots,
  filterSession,
}) => {
  return (
    <div className="md:hidden">
      {' '}
      {/* This div will only be visible on mobile screens */}
      {allTimeSlots.map((time, timeIndex) => (
        <div key={time} className="mb-4 border-b border-gray-300 pb-4">
          <h3 className="text-lg font-bold mb-2 font-roboto-mono text-black-02">
            {time}
          </h3>
          <div className="grid grid-cols-1 gap-2">
            {sessions
              .filter((session) => session.time_start === time)
              .filter((session) => filterSession(session, filters))
              .map((session) => (
                <div
                  key={session.id}
                  className="grid grid-cols-[60px_1fr] gap-2 items-center"
                >
                  <div
                    className={`h-full flex items-center justify-center text-center text-black-02 font-normal rounded-lg border-2 border-gray-800 ${getTrackColor(session.track)}`}
                  >
                    <span className="text-xs rotate-90 whitespace-nowrap">
                      {session.track}
                    </span>
                  </div>
                  <div>
                    <SessionCard session={session} isGrayedOut={false} />
                  </div>
                </div>
              ))}
          </div>
        </div>
      ))}
    </div>
  )
}

export default TimetableList
