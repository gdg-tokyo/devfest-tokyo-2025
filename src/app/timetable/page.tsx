'use client'

import OverlayMessageCard from '@/components/common/OverlayMessageCard'
import FilterSystem from '@/features/timetable/components/FilterSystem'
import TimetableGrid from '@/features/timetable/components/TimetableGrid'
import TimetableList from '@/features/timetable/components/TimetableList' // Import TimetableList
import { getSessions, getSpeakers, getTalks } from '@/lib/data-parser'
import { Session, Speaker, Talk } from '@/types'
import { useCallback, useEffect, useMemo, useState } from 'react'
import { generateTimeSlots, filterSession } from '@/features/timetable/utils' // Import from utils

const TimetablePage = () => {
  const allSessions: Session[] = getSessions()

  const [filters, setFilters] = useState<{ levels: string[]; keyword: string }>(
    { levels: [], keyword: '' }
  )
  const [showNoDataMessage, setShowNoDataMessage] = useState(false)

  useEffect(() => {
    if (allSessions.length === 0) {
      setShowNoDataMessage(true)
    } else {
      setShowNoDataMessage(false)
    }
  }, [allSessions])

  const handleFilterChange = useCallback(
    (newFilters: { levels: string[]; keyword: string }) => {
      setFilters(newFilters)
    },
    []
  )

  const allRawTalks = getTalks()
  const allSpeakers = getSpeakers()

  // --- Moved logic from TimetableGrid --- //
  const tracks = Array.from(
    new Set(allSessions.map((session) => session.track))
  ).sort()

  const displayTracks =
    tracks.length > 0 ? tracks : ['Track A', 'Track B', 'Track C', 'Track D']

  const allStartTimes = allSessions.map((s) => s.time_start)
  const allEndTimes = allSessions.map((s) => s.time_end)

  const earliestTime =
    allStartTimes.length > 0 ? allStartTimes.sort()[0] : '09:00'
  const latestTime =
    allEndTimes.length > 0 ? allEndTimes.sort().reverse()[0] : '18:00'

  const allTimeSlots = generateTimeSlots(allSessions)
  // --- End of moved logic --- //

  const hasMatchingSessions = useMemo(() => {
    // This logic needs to use the filterSession from utils.ts
    return allSessions.some((session) => filterSession(session, filters))
  }, [allSessions, filters])

  const availableLevels = useMemo(() => {
    const validLevels = ['Beginner', 'Intermediate', 'Advanced']
    const levels = allSessions
      .flatMap((s) => s.level || [])
      .filter((level) => validLevels.includes(level))
    return Array.from(new Set(levels))
  }, [allSessions])

  return (
    <div className="container mx-auto px-4 lg:px-8 max-w-screen-md lg:max-w-screen-xl p-4">
      <h1 className="text-3xl font-bold mb-6">Event Timetable</h1>
      <FilterSystem
        onFilterChange={handleFilterChange}
        availableLevels={availableLevels}
      />
      {allSessions.length > 0 ? (
        hasMatchingSessions ? (
          <>
            <TimetableList
              sessions={allSessions}
              filters={filters}
              allTimeSlots={allTimeSlots}
              filterSession={filterSession}
            />
            <TimetableGrid
              sessions={allSessions}
              filters={filters}
              allTimeSlots={allTimeSlots}
              displayTracks={displayTracks}
              filterSession={filterSession}
            />
          </>
        ) : (
          <p>No sessions available matching your criteria.</p>
        )
      ) : (
        <p>No sessions available.</p>
      )}
      <OverlayMessageCard
        message="Session data is not available. Please try again later."
        isVisible={showNoDataMessage}
      />
    </div>
  )
}

export default TimetablePage
