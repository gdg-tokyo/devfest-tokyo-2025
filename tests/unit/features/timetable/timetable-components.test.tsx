import React from 'react'
import { render, screen } from '@testing-library/react'
import '@testing-library/jest-dom'
import TimetableList from '@/features/timetable/components/TimetableList'
import TimetableGrid from '@/features/timetable/components/TimetableGrid'
import { Session } from '@/types'
import { getTrackColor } from '@/features/timetable/utils'

// Mock SessionCard to avoid deep rendering issues and focus on TimetableList/Grid structure
jest.mock('@/features/timetable/components/SessionCard', () => {
  // eslint-disable-next-line react/display-name
  return ({ session }: { session: Session }) => (
    <div data-testid={`session-card-${session.id}`}>{session.title}</div>
  )
})

// Mock matchMedia for responsive tests
const mockMatchMedia = (matches: boolean) => {
  Object.defineProperty(window, 'matchMedia', {
    writable: true,
    value: jest.fn().mockImplementation((query) => ({
      matches: matches,
      media: query,
      onchange: null,
      addListener: jest.fn(), // Deprecated
      removeListener: jest.fn(), // Deprecated
      addEventListener: jest.fn(),
      removeEventListener: jest.fn(),
      dispatchEvent: jest.fn(),
    })),
  })
}

describe('TimetableList (Mobile View)', () => {
  const mockSessions: Session[] = [
    {
      id: 's1',
      title: 'Mobile Session 1',
      time_start: '09:00',
      time_end: '09:30',
      track: 'Track A',
      level: ['Beginner'],
      talk_ids: [],
      description: 'Description 1',
    },
    {
      id: 's2',
      title: 'Mobile Session 2',
      time_start: '09:00',
      time_end: '09:30',
      track: 'Track B',
      level: ['Intermediate'],
      talk_ids: [],
      description: 'Description 2',
    },
    {
      id: 's3',
      title: 'Mobile Session 3',
      time_start: '10:00',
      time_end: '10:30',
      track: 'Track A',
      level: ['Advanced'],
      talk_ids: [],
      description: 'Description 3',
    },
  ]

  const mockFilters = { levels: [], keyword: '' }
  const mockAllTimeSlots = ['09:00', '10:00']
  const mockFilterSession = jest.fn(() => true)

  beforeEach(() => {
    mockMatchMedia(true) // Simulate mobile screen
  })

  it('renders correctly on mobile screens', () => {
    const { asFragment } = render(
      <TimetableList
        sessions={mockSessions}
        filters={mockFilters}
        allTimeSlots={mockAllTimeSlots}
        filterSession={mockFilterSession}
      />
    )
    expect(asFragment()).toMatchSnapshot()
  })

  it('displays sessions stacked vertically with track name panel', () => {
    render(
      <TimetableList
        sessions={mockSessions}
        filters={mockFilters}
        allTimeSlots={mockAllTimeSlots}
        filterSession={mockFilterSession}
      />
    )

    const timeSlots = screen.getAllByRole('heading', { level: 3 })
    expect(timeSlots).toHaveLength(mockAllTimeSlots.length)

    mockSessions.forEach((session) => {
      expect(screen.getByText(session.title)).toBeInTheDocument()
      expect(screen.getByText(session.title)).toBeInTheDocument()

      // Find the parent div that contains both the track panel and the session card
      const sessionEntryDiv = screen.getByTestId(`session-card-${session.id}`)
        .parentElement?.parentElement
      expect(sessionEntryDiv).toBeInTheDocument()

      // Now, within this sessionEntryDiv, find the track panel (first child div)
      const trackPanel = sessionEntryDiv?.querySelector('div:first-child')
      expect(trackPanel).toBeInTheDocument()

      // And within the trackPanel, find the span with the track name
      const trackNameSpan = trackPanel?.querySelector(
        'span.text-xs.rotate-90.whitespace-nowrap'
      )
      expect(trackNameSpan).toBeInTheDocument()
      expect(trackNameSpan).toHaveTextContent(session.track)

      expect(trackPanel).toHaveClass(getTrackColor(session.track))
      expect(trackPanel).toHaveClass('border-2')
      expect(trackPanel).toHaveClass('border-gray-800')
      expect(trackPanel).toHaveClass('text-black-02')
      expect(trackPanel).toHaveClass('font-normal')
      expect(trackPanel).toHaveClass(getTrackColor(session.track))
      expect(trackPanel).toHaveClass('border-2')
      expect(trackPanel).toHaveClass('border-gray-800')
      expect(trackPanel).toHaveClass('text-black-02')
      expect(trackPanel).toHaveClass('font-normal')
    })
  })
})

describe('TimetableGrid (Tablet/PC View)', () => {
  const mockSessions: Session[] = [
    {
      id: 's1',
      title: 'Grid Session 1',
      time_start: '09:00',
      time_end: '09:30',
      track: 'Track A',
      level: ['Beginner'],
      talk_ids: [],
      description: 'Description 1',
    },
    {
      id: 's2',
      title: 'Grid Session 2',
      time_start: '09:00',
      time_end: '09:30',
      track: 'Track B',
      level: ['Intermediate'],
      talk_ids: [],
      description: 'Description 2',
    },
  ]

  const mockFilters = { levels: [], keyword: '' }
  const mockAllTimeSlots = ['09:00', '09:30']
  const mockDisplayTracks = ['Track A', 'Track B']
  const mockFilterSession = jest.fn(() => true)

  beforeEach(() => {
    mockMatchMedia(false) // Simulate non-mobile screen
  })

  it('renders correctly on tablet/PC screens', () => {
    const { asFragment } = render(
      <TimetableGrid
        sessions={mockSessions}
        filters={mockFilters}
        allTimeSlots={mockAllTimeSlots}
        displayTracks={mockDisplayTracks}
        filterSession={mockFilterSession}
      />
    )
    expect(asFragment()).toMatchSnapshot()
  })

  it('displays sessions in a grid layout with track headers', () => {
    render(
      <TimetableGrid
        sessions={mockSessions}
        filters={mockFilters}
        allTimeSlots={mockAllTimeSlots}
        displayTracks={mockDisplayTracks}
        filterSession={mockFilterSession}
      />
    )

    mockDisplayTracks.forEach((track) => {
      expect(screen.getByText(track)).toBeInTheDocument()
    })

    mockSessions.forEach((session) => {
      expect(screen.getByText(session.title)).toBeInTheDocument()
    })
  })
})
