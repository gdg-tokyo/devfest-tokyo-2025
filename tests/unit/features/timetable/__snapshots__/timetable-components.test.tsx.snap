// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`TimetableGrid (Tablet/PC View) renders correctly on tablet/PC screens 1`] = `
<DocumentFragment>
  <div
    class="overflow-x-auto hidden md:block"
  >
     
    <div
      class="jsx-3422117831 grid gap-2 p-2"
      style="grid-template-rows: [tracks] auto repeat(2, minmax(40px, auto));"
    >
      <div
        class="jsx-3422117831 sticky left-0 top-0 p-3 z-10 subhead text-sm text-gray-600"
      >
        Time
      </div>
      <div
        class="jsx-3422117831 p-3 rounded-lg border-2 border-gray-800 text-center sticky top-0 z-10 bg-gdg-pastel-red"
        style="grid-column: 2; grid-row: 1;"
      >
        <h5
          class="jsx-3422117831"
        >
          Track A
        </h5>
      </div>
      <div
        class="jsx-3422117831 p-3 rounded-lg border-2 border-gray-800 text-center sticky top-0 z-10 bg-gdg-pastel-blue"
        style="grid-column: 3; grid-row: 1;"
      >
        <h5
          class="jsx-3422117831"
        >
          Track B
        </h5>
      </div>
      <div
        class="jsx-3422117831 sticky left-0 p-2 border-r border-gray-300 font-roboto-mono text-sm text-black-02 flex items-center justify-center z-10"
        style="grid-row: 2; grid-column: 1;"
      >
        09:00
      </div>
      <div
        class="jsx-3422117831 sticky left-0 p-2 border-r border-gray-300 font-roboto-mono text-sm text-black-02 flex items-center justify-center z-10"
        style="grid-row: 3; grid-column: 1;"
      >
        09:30
      </div>
      <div
        class="jsx-3422117831 font-google-sans text-black-02 "
        style="grid-row: 2 / span 1; grid-column: 2; z-index: 5;"
      >
        <div
          data-testid="session-card-s1"
        >
          Grid Session 1
        </div>
      </div>
      <div
        class="jsx-3422117831 font-google-sans text-black-02 "
        style="grid-row: 2 / span 1; grid-column: 3; z-index: 5;"
      >
        <div
          data-testid="session-card-s2"
        >
          Grid Session 2
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`TimetableList (Mobile View) renders correctly on mobile screens 1`] = `
<DocumentFragment>
  <div
    class="md:hidden"
  >
     
    <div
      class="mb-4 border-b border-gray-300 pb-4"
    >
      <h3
        class="text-lg font-bold mb-2 font-roboto-mono text-black-02"
      >
        09:00
      </h3>
      <div
        class="grid grid-cols-1 gap-2"
      >
        <div
          class="grid grid-cols-[60px_1fr] gap-2 items-center"
        >
          <div
            class="h-full flex items-center justify-center text-center text-black-02 font-normal rounded-lg border-2 border-gray-800 bg-gdg-pastel-red"
          >
            <span
              class="text-xs rotate-90 whitespace-nowrap"
            >
              Track A
            </span>
          </div>
          <div>
            <div
              data-testid="session-card-s1"
            >
              Mobile Session 1
            </div>
          </div>
        </div>
        <div
          class="grid grid-cols-[60px_1fr] gap-2 items-center"
        >
          <div
            class="h-full flex items-center justify-center text-center text-black-02 font-normal rounded-lg border-2 border-gray-800 bg-gdg-pastel-blue"
          >
            <span
              class="text-xs rotate-90 whitespace-nowrap"
            >
              Track B
            </span>
          </div>
          <div>
            <div
              data-testid="session-card-s2"
            >
              Mobile Session 2
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="mb-4 border-b border-gray-300 pb-4"
    >
      <h3
        class="text-lg font-bold mb-2 font-roboto-mono text-black-02"
      >
        09:30
      </h3>
      <div
        class="grid grid-cols-1 gap-2"
      />
    </div>
    <div
      class="mb-4 border-b border-gray-300 pb-4"
    >
      <h3
        class="text-lg font-bold mb-2 font-roboto-mono text-black-02"
      >
        10:00
      </h3>
      <div
        class="grid grid-cols-1 gap-2"
      >
        <div
          class="grid grid-cols-[60px_1fr] gap-2 items-center"
        >
          <div
            class="h-full flex items-center justify-center text-center text-black-02 font-normal rounded-lg border-2 border-gray-800 bg-gdg-pastel-red"
          >
            <span
              class="text-xs rotate-90 whitespace-nowrap"
            >
              Track A
            </span>
          </div>
          <div>
            <div
              data-testid="session-card-s3"
            >
              Mobile Session 3
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
